### PR TITLE
Added spin loop pause function

### PR DESCRIPTION
**Pull request by mstewartgallus**
This is the one function I wish was added to the C11 atomics support and I feel would be really convenient in Rust.

### CODE REVIEW

File: src/libstd/sync/mod.rs [hunk 1]
```diff
@@ -42,3 +42,18 @@ mod condvar;
...
+/// Save power or switch hyperthreads in a busy-wait spin-loop.  This
+/// function is deliberately more primitive than
+/// `std::thread::yield_now` and does not directly yield to the
+/// system's scheduler.  In some cases it might be useful to use a
+/// combination of both functions.
+#[inline]
+#[stable(feature = "rust1", since = "1.1.6")]
+pub fn spin_loop_pause()
+{
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
```
Comment (sfackler) [inline src/libstd/sync/mod.rs]: We should probably support more than just x86 if this were to merge - is there some LLVM intrinsic we could leverage here?

**Comment by sfackler** [comment]
cc @rust-lang/libs thoughts?

File: src/libstd/sync/mod.rs [hunk 2]
```diff
@@ -42,3 +42,18 @@ mod condvar;
...
 mod once;
 mod rwlock;
+
+/// Save power or switch hyperthreads in a busy-wait spin-loop.  This
+/// function is deliberately more primitive than
+/// `std::thread::yield_now` and does not directly yield to the
+/// system's scheduler.  In some cases it might be useful to use a
+/// combination of both functions.
+#[inline]
+#[stable(feature = "rust1", since = "1.1.6")]
```
Comment (sfackler) [inline src/libstd/sync/mod.rs]: stable?

File: src/libstd/sync/mod.rs (hunk 1 above)
Comment (BurntSushi) [inline src/libstd/sync/mod.rs]: [`_mm_pause`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=pause&expand=3864) is part of SSE2, so we should probably use that. But that is exactly as portable as what's provided here I think. (If `_mm_pause` is the answer, then we should probably punt on this until SIMD.)

**Comment by alexcrichton** [comment]
If this was explicitly left out of a c11 memory model or something like that then I'd personally not be in favor of such a function. 

@sstewartgallus do you have more motivation for adding this beyond "I feel would be really convenient in Rust" to add this?

**Comment by sstewartgallus** [comment]
The motivating case is that every lock-free data-structure in crossbeam should use this function. IMO, almost every time you use atomics you should use this function. Secondly, this function causes no incompatibility problems with platforms it doesn't work for and is just a nop on those.

File: src/libstd/sync/mod.rs (hunk 2 above)
Comment (sstewartgallus) [inline src/libstd/sync/mod.rs]: I'm fine if this isn't stable.

**Comment by sstewartgallus** [comment]
See openjdk.java.net/jeps/285  for previous work.

**Comment by Amanieu** [comment]
https://github.com/Amanieu/parking_lot/blob/master/core/src/spinwait.rs#L41

**Comment by Amanieu** [comment]
This doesn't really have to be in the stdlib, it could just be a library on crates.io. However we do need stable inline asm for this to be widely used.

File: src/libstd/sync/mod.rs (hunk 2 above)
Comment (sstewartgallus) [inline src/libstd/sync/mod.rs]: @sfackler Actually can you help me out with the stability attribute syntax? I can't find any reference of it online.

Comment (sfackler) [inline src/libstd/sync/mod.rs]: Here's an example: https://github.com/rust-lang/rust/blob/master/src/libcore/convert.rs#L291

Pick some arbitrary (unique) feature name, and before we merge this we'll create a tracking issue to refer to.

**Comment by sfackler** [comment]
I think it probably makes sense to land this unstable for now at least.

@rfcbot fcp merge

**Comment by alexcrichton** [comment]
The documentation here seems particularly lacking to me I think? There's no dicussion of cross-platform concerns, portability, etc. I'd personally prefer to see at least a warning that this function isn't guaranteed to do anything, it's just a small optimization on some architectures in some situations. I also feel like this'd best be inside of libcore, not inside of libstd. Finally, I'm not personally sold on the `std::sync::spin_loop_pause` method. This sort of feels like it belongs in the `atomic` module and would want a name that doesn't necessarily have "spin loop" in the name. 

Looks like tidy is also failing?

**Pull request closed without merging**